connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java

```diff
@@ -254,7 +260,12 @@ public Map<ByteBuffer, ByteBuffer> get(long timeout, TimeUnit unit) throws Inter
      * write to that store, and the passed-in {@link Callback} is invoked once that write completes. If a worker-global
      * store is provided, a secondary write is made to that store if the write to the connector-specific store
      * succeeds. Errors with this secondary write are not reflected in the returned {@link Future} or the passed-in
-     * {@link Callback}; they are only logged as a warning to users.
+     * {@link Callback}; they are only logged as a warning to users. The only exception to this rule is when the offsets
+     * that need to be committed contains tombstone records as well. In such cases, a write consisting of only tombstone
+     * offsets would first happen on the worker-global store and only if it succeeds, would all the offsets be written
+     * to the connector-specific store and the regular offsets would be written to the worker-global store. Note that
+     * in this case, failure to write regular offsets to secondary store would still not reflect in the returned
+     * {@link Future} or the passed-in {@link Callback}
```

Comment:
Some suggestions on making this clearer and easier to read (begins at line 262, which I can't include in a suggestion block in GitHub):
```suggestion
     * succeeds.
     * <p>
     * Normally, errors with this secondary write are not reflected in the returned {@link Future} or the passed-in
     * {@link Callback}; they are only logged as a warning to users. The only exception to this rule is when the
     * offsets that need to be committed contain tombstone records.
     * <p>When the to-be-committed offsets contain tombstones, offset commits take place in three phases:
     * <ol>
     *     <li>First, only the tombstone offsets are written to the worker-global store. Failures during this step will
     *     be reflected in the returned {@link Future} and reported to the passed-in {@link Callback}.</li>
     *     <li>If and only if the previous write to the worker-global store succeeded, all offsets (both tombstones and
     *     non-tombstones) are written to the connector-specific store. Failures during this step will also be
     *     reflected in the returned {@link Future} and reported to the passed-in {@link Callback}.</li>
     *     <li>Finally, if and only if the previous write to the connector-specific store succeeded, all offsets with
     *     non-tombstone values are written to the worker-global store. Failures during this step will only be reported
     *     as warning log messages, and will not be reflected in the returned {@link Future} or reported to the
     *     passed-in {@link Callback}.</li>
     * </ol>
```

Reply:
Added.